src/python_ml_project_template/utils/script_utils.py

```diff
@@ -60,9 +45,51 @@ def flatten_outputs(outputs: List[TorchTree]) -> TorchTree:
     return cast(TorchTree, output_dict)
 
 
+class CanMakePlots(Protocol):
+    @staticmethod
+    @abc.abstractmethod
+    def make_plots(preds, batch: tgd.Batch):
+        pass
+
+
+class LightningModuleWithPlots(pl.LightningModule, CanMakePlots):
+    pass
+
+
+# TODO: Change
 class LogPredictionSamplesCallback(Callback):
-    def __init__(self, logger: WandbLogger):
+    def __init__(self, logger: WandbLogger, eval_per_n_epoch):
         self.logger = logger
+        self.eval_per_n_epoch = eval_per_n_epoch
+
+    @staticmethod
+    def eval_log_random_sample(
+        trainer: pl.Trainer,
+        pl_module: LightningModuleWithPlots,
+        batch,
+        prefix: Literal["train", "val", "unseen"],
+    ):
+        # preds = outputs[np.random.randint(0, outputs.shape[0])]
+        data = batch[np.random.randint(0, len(batch))]
+        data = tgd.Batch.from_data_list([data]).to(pl_module.device)
+        with torch.no_grad():
+            pl_module.eval()
+            preds = pl_module(data)
```

Comment:
you probably don't have to re-evaluate here. because "outputs" is an argument to `on_validation_batch_end`, so you should already have the batch and the predictions. also you still need:

```
if batch_idx == 0:
```

so you only make plots on the first batch of each validation epoch.